geo/src/algorithm/bounding_rect.rs

```diff
@@ -0,0 +1,290 @@
+use {
+    Coordinate, CoordinateType, Line, LineString, MultiLineString, MultiPoint, MultiPolygon,
+    Polygon, Rect, Triangle,
+};
+
+/// Calculation of the bounding rectangle of a geometry.
+pub trait BoundingRect<T: CoordinateType> {
+    type Output;
+
+    /// Return the bounding rectangle of a geometry
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use geo::{Point, LineString};
+    /// use geo::algorithm::bounding_rect::BoundingRect;
+    ///
+    /// let mut vec = Vec::new();
+    /// vec.push(Point::new(40.02f64, 116.34));
+    /// vec.push(Point::new(42.02f64, 116.34));
+    /// vec.push(Point::new(42.02f64, 118.34));
+    /// let linestring = LineString::from(vec);
+    /// let bounding_rect = linestring.bounding_rect().unwrap();
+    ///
+    /// assert_eq!(40.02f64, bounding_rect.min.x);
+    /// assert_eq!(42.02f64, bounding_rect.max.x);
+    /// assert_eq!(116.34, bounding_rect.min.y);
+    /// assert_eq!(118.34, bounding_rect.max.y);
+    /// ```
+    ///
+    fn bounding_rect(&self) -> Self::Output;
+}
+
+fn get_min_max<T>(p: T, min: T, max: T) -> (T, T)
+where
+    T: CoordinateType,
+{
+    if p > max {
+        (min, p)
+    } else if p < min {
+        (p, max)
+    } else {
+        (min, max)
+    }
+}
+
+fn get_bounding_rect<I, T>(collection: I) -> Option<Rect<T>>
+where
+    T: CoordinateType,
+    I: IntoIterator<Item = Coordinate<T>>,
+{
+    let mut iter = collection.into_iter();
+    if let Some(pnt) = iter.next() {
+        let mut xrange = (pnt.x, pnt.x);
+        let mut yrange = (pnt.y, pnt.y);
+        for pnt in iter {
+            let (px, py) = pnt.x_y();
+            xrange = get_min_max(px, xrange.0, xrange.1);
+            yrange = get_min_max(py, yrange.0, yrange.1);
+        }
+        return Some(Rect {
+            min: Coordinate {
+                x: xrange.0,
+                y: yrange.0,
+            },
+            max: Coordinate {
+                x: xrange.1,
+                y: yrange.1,
+            },
+        });
+    }
+    None
+}
+
+impl<T> BoundingRect<T> for MultiPoint<T>
+where
+    T: CoordinateType,
+{
+    type Output = Option<Rect<T>>;
+
+    ///
+    /// Return the BoundingRect for a MultiPoint
+    ///
+    fn bounding_rect(&self) -> Self::Output {
+        get_bounding_rect(self.0.iter().map(|p| p.0))
+    }
+}
+
+impl<T> BoundingRect<T> for Line<T>
```

Comment:
Are the Line and LineString implementations intentionally duplicated in geo-types/src/algorithms.rs?

Reply:
yeah, intentional. to have `spade` trait implementations for the types in `geo-types`, we need to have some basic algorithms included in the `geo-types` crate. but i'm planning to consolidate them in a follow-up pull request to this one.